### ISSUE TITLE
Some questions about Memory

### ISSUE THREAD

**Issue by 9p15p** [seoungwugoh/STM#14]
Hi, sir! Thank you for your fine work, but I still have some question.

1. In the "eval_DAVIS.py", the memory is saved at every 5 frames. But in the train, We only use 3 frames in each iteration. **How should we save the memory in the train?**

2.Should we make train_code's memory part(Sth like memory part in eval_code) **under "with torch.no_grad"?**Does memory require grad?

**Comment by seoungwugoh** [comment]
@9p15p Yes, memory also requires gradients. In specific, we learn how to encode a memory from [frame, mask]. Naturally, every feed-forward operation is done *without* `torch.no_grad` during training.

**Comment by 9p15p** [comment]
Thank you for your reply.
I have another question: how should we backpropagate our loss ?IOW, where should we put our  "loss.backward"?
##I train the model by only using a object with 3 frames and 3masks at a time ##

1. calculate loss and put "loss.backward"  in every object every frame.(in experiment,  we will use it with 'retain_graph=True'  twice for the last two frames and after we calculate all frames' loss we use another one **without** 'retain_graph=True'.

2. calculate loss in every object every frame but only put "loss.backward" after we calculate all frames' loss.

**Comment by seoungwugoh** [comment]
What I did is option 2. 
We sum-up all the losses and call backward and step at the end of the iteration.

**Comment by 9p15p** [comment]
Thank you, Sir. 
Although,my best reimplemention is still not satisfactory.
But in my own experiment, calling loss back every frame has a better performance and Converges quicker. （use  'retain_graph=True' for the second frame's loss.）.I will try again.


best wishes!

**Comment by ftlong6666** [comment]
Hello @seoungwugoh I stuck in this error.
```
File "E:\FGVC\STM\model.py", line 219, in memorize
    B_list['m'].append(masks[:,o])
IndexError: index 11 is out of bounds for dimension 1 with size 11
```

what is the meaning of `masks[:,o]`? I tried to find the `num_objects` it returns this.
`tensor([255], device='cuda:0')`
